### PR TITLE
Apply Rector suggestions for PHP 8.1 

### PR DESCRIPTION
**Pull request by asumaran**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Subset of WPCOM changes for PHP 8.1 compatibility based on the D128937-code diff.

Ref: p7H4VZ-4x1-p2

### How to test the changes in this Pull Request:

The changes on this PR are based on changes currently made in WPCOM, As there are no logic changes, only compatibility with PHP 8.1, it was deemed unnecessary to include testing instructions. To put it simply, the current tests should be able to complete without any problems, and that should be sufficient.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Review by prettyboymp — COMMENTED** [review]
changes look good.

**Review by vedanshujain — COMMENTED** [review]
There are linting errors that needs to be fixed: https://github.com/woocommerce/woocommerce/actions/runs/7561652109/job/20590414324?pr=43236. Otherwise looks good, I can test changes for order table data stores, but we'd need to have changes in MarketingSpecs and ProductForm/Componenet as well.

**Comment by asumaran** [comment]
@vedanshujain I've fixed the Lint errors https://github.com/woocommerce/woocommerce/pull/43236/commits/41349cf333c3ad61b5025a54a05786762a8ff029.  Please take another look at this.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableMetaQuery.php [hunk 1]
```diff
@@ -541,7 +544,7 @@ private function generate_where_for_clause_key( array $clause ): string {
 				$where              = $wpdb->prepare( "$alias.meta_key LIKE %s", $meta_compare_value ); // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
 				break;
 			case 'IN':
-				$meta_compare_string = "$alias.meta_key IN (" . substr( str_repeat( ',%s', count( $clause['key'] ) ), 1 ) . ')';
+				$meta_compare_string = "$alias.meta_key IN (" . substr( str_repeat( ',%s', count( (array) $clause['key'] ) ), 1 ) . ')';
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableMetaQuery.php:547]: to test

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php [hunk 2]
```diff
@@ -292,11 +292,10 @@ private function generate_where_for_clause( $clause ): string {
 		}
 
 		$clause_compare = $clause['compare'];
-
 		switch ( $clause_compare ) {
 			case 'IN':
 			case 'NOT IN':
-				$where = $wpdb->prepare( '(' . substr( str_repeat( ',%s', count( $clause_value ) ), 1 ) . ')', $clause_value ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+				$where = $wpdb->prepare( '(' . substr( str_repeat( ',%s', count( (array) $clause_value ) ), 1 ) . ')', $clause_value ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableFieldQuery.php:298]: to test.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableMetaQuery.php [hunk 3]
```diff
@@ -566,7 +569,7 @@ private function generate_where_for_clause_key( array $clause ): string {
 				$where              = $wpdb->prepare( $meta_compare_string, $meta_compare_value ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
 				break;
 			case 'NOT IN':
-				$array_subclause     = '(' . substr( str_repeat( ',%s', count( $clause['key'] ) ), 1 ) . ') ';
+				$array_subclause     = '(' . substr( str_repeat( ',%s', count( (array) $clause['key'] ) ), 1 ) . ') ';
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableMetaQuery.php:572]: to test

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableMetaQuery.php [hunk 4]
```diff
@@ -618,7 +621,7 @@ private function generate_where_for_clause_value( $clause ): string {
 		switch ( $meta_compare ) {
 			case 'IN':
 			case 'NOT IN':
-				$where = $wpdb->prepare( '(' . substr( str_repeat( ',%s', count( $meta_value ) ), 1 ) . ')', $meta_value ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+				$where = $wpdb->prepare( '(' . substr( str_repeat( ',%s', count( (array) $meta_value ) ), 1 ) . ')', $meta_value ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableMetaQuery.php:624]: to test

**Review by vedanshujain — COMMENTED** [review]
I have marked some places that are not covered by unit tests looks like and needs manual testing. Planning to come back to those and test it, otherwise looks good.

**Review by vedanshujain — APPROVED** [review]
LGTM! I have few unit tests around untested paths.

**Pull request merged**